tuned/exports/unix_socket_exporter.py

```diff
@@ -0,0 +1,228 @@
+		if method_name in self._unix_socket_methods:
+			raise Exception("Method with this name (%s) is already exported." % method_name)
+
+		class wrapper(object):
+			def __init__(self, in_signature, out_signature):
+				self._in_signature = in_signature
+				self._out_signature = out_signature
+				
+			def __call__(self, *args, **kwargs):
+				if len(args) + len(kwargs) != len(self._in_signature):
+					# TODO: Add type check for parameters
+					raise Exception("Wrong count of params for method '%s'." % method_name)
+				return method(*args, **kwargs)
+
+		self._unix_socket_methods[method_name] = wrapper(in_signature, out_signature)
+
+	def signal(self, method, out_signature):
+		if not ismethod(method):
+			raise Exception("Only bound methods can be exported.")
+		
+		method_name = method.__name__
+		if method_name in self._unix_socket_methods:
+			raise Exception("Method with this name (%s) is already exported." % method_name)
+		
+		class wrapper(object):
+			def __init__(self, out_signature):
+				self._out_signature = out_signature
+			
+			def __call__(self, *args, **kwargs):
+				return method(*args, **kwargs)
+		
+		self._unix_socket_methods[method_name] = wrapper(out_signature)
+		self._signals.add(method_name)
+
+	def send_signal(self, signal, *args, **kwargs):
+		if not signal in self._signals:
+			raise Exception("Signal '%s' doesn't exist." % signal)
+		s = socket.socket(socket.AF_UNIX, socket.SOCK_DGRAM)
+		s.setblocking(False)
+		for p in self._socket_signal_paths:
+			try:
+				self._send_data(s, p, "sig", list(args) + list(kwargs.values()))
+			except OSError as e:
+				log.warning("Error while sending signal '%s' to socket '%s': %s" % (signal, p, e))
+
+	def register_signal_path(self, path):
+		self._socket_signal_paths.append(path)
+
+	def _construct_socket_object(self):
+		if self._socket_path:
+			if os.path.exists(self._socket_path):
+				os.unlink(self._socket_path)
+			s = socket.socket(socket.AF_UNIX, socket.SOCK_DGRAM)
+			s.setblocking(False)
+			s.bind(self._socket_path)
+			os.chown(self._socket_path, self._ownership[0], self._ownership[1])
+			if self._permissions:
+				os.chmod(self._socket_path, self._permissions)
+			self._socket_object = s
+
+	def start(self):
+		if self.running():
+			return
+
+		self.stop()
+		self._construct_socket_object()
+
+	def stop(self):
+		if self._socket_object:
+			self._socket_object.close()
+
+	def _send_data(self, s, address, type, data):
+		"""
+		Sends data via socket s to address.
+		First byte of message specifies type of message (currently "r" for response, "e" for error message and "s" for signal).
+		Four more bytes are integer in network/big-endian with length of rest of message.
+		Rest of message is json data.
+		"""
+		log.debug("Sending socket data (type: %s, data: %s) to %s" % (type, data, address))
+		try:
+			j = json.dumps(data).encode("utf-8")
+			s.sendto(struct.pack("!sL", type.encode("utf-8"), len(j)), address)
+			s.sendto(j, address)
```

Comment:
Why two datagrams here? JSON does not need extra length, moreover recv is reading just one datagram per call and the buff size is the maximum received, extra data will be dropped (as this is datagram not stream).

So you really only need the json one.

Reply:
See here for an example: https://gist.github.com/MarSik/e68acb69611d23939c835bc2dd6aa85d